src/test/index.ts

```diff
@@ -1,11 +1,62 @@
 // This file is providing the test runner to use when running extension tests.
-import * as testRunner from 'vscode/lib/testrunner';
+import * as path from 'path';
+import * as vscode from 'vscode';
+import glob = require('glob');
+import Mocha = require('mocha');
 
-// You can directly control Mocha options by uncommenting the following lines
-// See https://github.com/mochajs/mocha/wiki/Using-mocha-programmatically#set-options for more info
-testRunner.configure({
-	ui: 'bdd', 		// the BDD UI is being used in extension.test.ts (describe, it, etc.)
-	useColors: true // colored output from test results
-});
+import { mockWebviewEnvironment } from './mocks/mockWebviewEnvironment';
 
-module.exports = testRunner;
\ No newline at end of file
+// Linux: prevent a weird NPE when mocha on Linux requires the window size from the TTY
+// Since we are not running in a tty environment, we just implement the method statically.
+// This is copied verbatim from the upstream, default Mocha test runner:
+// https://github.com/microsoft/vscode-extension-vscode/blob/master/lib/testrunner.ts
+const tty = require('tty') as any;
+if (!tty.getWindowSize) {
+	tty.getWindowSize = function () { return [80, 75]; };
+}
+
+function addTests(mocha: Mocha, root: string): Promise<void> {
+	return new Promise((resolve, reject) => {
+		glob('**/**.test.js', {cwd: root}, (error, files) => {
+			if (error) {
+				return reject(error);
+			}
+
+			for (const testFile of files) {
+				mocha.addFile(path.join(root, testFile));
+			}
+			resolve();
+		});
+	});
+}
+
+async function runAllExtensionTests(testsRoot: string): Promise<number> {
+	// Ensure the dev-mode extension is activated
+	const { name, publisher } = require('../../package.json') as { name: string, publisher: string };
+	const extensionId = `${publisher}.${name}`;
+	await vscode.extensions.getExtension(extensionId)!.activate();
+
+	mockWebviewEnvironment.install(global);
+
+	const mocha = new Mocha({
+		ui: 'bdd',
+		useColors: true,
+	});
+
+	await addTests(mocha, testsRoot);
+	await addTests(mocha, path.resolve(testsRoot, '../../preview-src/dist/preview-src/test'));
```

Comment:
This needs to be a custom runner primarily so that we can add tests from `preview-src/`. It's also a convenient place to install the `MockWebviewEnvironment`.